test/exceptions.jl

```diff
@@ -0,0 +1,210 @@
+using Test
+
+@testset "Exception stack nesting" begin
+    # Basic exception stack handling
+    try
+        error("A")
+    catch
+        @test length(catch_stack()) == 1
+    end
+    @test length(catch_stack()) == 0
+    try
+        try
+            error("A")
+        finally
+            @test length(catch_stack()) == 1
+        end
+    catch
+        @test length(catch_stack()) == 1
+    end
+    # Errors stack up
+    try
+        error("RootCause")
+    catch
+        @test length(catch_stack()) == 1
+        try
+            error("B")
+        catch
+            stack = catch_stack()
+            @test length(stack) == 2
+            @test stack[1][1].msg == "RootCause"
+            @test stack[2][1].msg == "B"
+        end
+        # Stack pops correctly
+        stack = catch_stack()
+        @test length(stack) == 1
+        @test stack[1][1].msg == "RootCause"
+    end
+    # Lowering - value position
+    val = try
+        error("A")
+    catch
+        @test length(catch_stack()) == 1
+        1
+    end
+    @test val == 1
+    function test_exc_stack_tailpos()
+        # exercise lowering code path for tail position
+        try
+            error("A")
+        catch
+            length(catch_stack())
+        end
+    end
+    @test test_exc_stack_tailpos() == 1
+    @test length(catch_stack()) == 0
+end
+
+@testset "Exception stacks and gotos" begin
+    function test_exc_stack_catch_return()
+        try
+            error("A")
+        catch
+            @test length(catch_stack()) == 1
+            return
+        end
+    end
```

Comment:
Is the following form also handled:
```julia
         try
             return
         catch
         finally
         end
```

Reply:
Should be handled currently, but I'll add a couple of extra test cases for the future.

To explain - currently these three block
```julia
try
    A
catch
    B
finally
    C
end
```

are lowered into the nested form

```julia
try
    try
        A
    catch
        B
    end
finally
    C
end
```
very early on which is nice and simple but also rather inefficient because we could get away with a single eh frame here. Something for future work perhaps ;-)

Reply:
Or if you're asking about the early-return-from-try form — if you're in the try block there's no exception popping required. I'll add a few tests for that too.